PEP 770: Add the dist-info.files table and reserve dist-info dirs

sethmlarson opened the pull request:
<!--
**Please** read our Contributing Guidelines (CONTRIBUTING.rst)
to make sure this repo is the right place for your proposed change. Thanks!
-->

* Change is either:
    * [x] To a Draft PEP
    * [ ] To an Accepted or Final PEP, with Steering Council approval
    * [ ] To fix an editorial issue (markup, typo, link, header, etc)
* [x] PR title prefixed with PEP number (e.g. ``PEP 123: Summary of changes``)

cc @rgommers 

<!-- readthedocs-preview pep-previews start -->
----
📚 Documentation preview 📚: https://pep-previews--4283.org.readthedocs.build/

<!-- readthedocs-preview pep-previews end -->

jkowalleck commented:
I believe this proposal constitutes a significant scope change and is not aligned with the original objectives.

The initial scope was to add SBOM to packages. However, the proposed changes introduce the addition of arbitrary data to packages, which significantly broadens the scope. This new capability/topic deviates from the original focus and introduces complexities that will progress at a different pace than the initial scope. It appears that this change could be an attempt to shift the focus or slow down the progress by diverting from the original scope.

I strongly oppose this PR and the proposed changes. If there is a substantial interest in including arbitrary data in packages, a separate PEP should be initiated.

rgommers commented:
@jkowalleck please don't post the same comment in two places. You seem to be new to Python packaging conversations, so for context: Discourse is the right place for high-level discussion and comments like "I prefer approach X over approach Y because Z"; PEP PRs are more editorial and for initial reviews on correctness/completeness before posting to a broader audience on Discourse.

rgommers commented:
@sethmlarson the description of PEP 725 looks pretty good as well. I'd suggest one small tweak to make the first bullet point more clear, since "abstract dependencies" isn't a well-defined term and you're only giving `virtual:` rather than a `pkg:` example. I suggest to add a more apples to apples comparison for OpenSSL/libssl: the PEP 725 dependency specifier `pkg:generic/openssl` is what will result in `pkg:pkg:rpm/almalinux/libssl3@3.2.0` in an SBOM.

The analogy that I tend to use is "dependencies in `pyproject.toml` vs. "resolved dependencies in a lock file".

jkowalleck commented:
re: https://github.com/python/peps/pull/4283#pullrequestreview-2643691363
> You may want to mention somewhere that using a separate `[additional-files]` key means that packages can start using it directly after acceptance of this PEP, rather than have a >1 yr rollout plan that would have been needed with a core metadata version bump.

Is "additional-files" part of [core metadata](https://packaging.python.org/en/latest/specifications/core-metadata/)?

pfmoore commented:
> Is "additional-files" part of [core metadata](https://packaging.python.org/en/latest/specifications/core-metadata/) already?

It's not a metadata item, it's a key in `pyproject.toml`.

jkowalleck commented:
> re: [#4283 (review)](https://github.com/python/peps/pull/4283#pullrequestreview-2643691363)
> 
> > You may want to mention somewhere that using a separate `[additional-files]` key means that packages can start using it directly after acceptance of this PEP, rather than have a >1 yr rollout plan that would have been needed with a core metadata version bump.
> 
> Is "additional-files" part of [core metadata](https://packaging.python.org/en/latest/specifications/core-metadata/) already?





> > Is "additional-files" part of [core metadata](https://packaging.python.org/en/latest/specifications/core-metadata/) already?
> 
> It's not a metadata item, it's a key in `pyproject.toml`.

Could you explain what this "rather than have a >1 yr rollout plan that would have been needed with a core metadata version bump" is supposed to mean?

pfmoore reviewed (approved):
This looks great to me. I think any comments I might have had have already been covered by @rgommers

You've put a lot more details into the specification of `additional-files` than I had anticipated, but that's a good thing - it feels like it's well prepared for any future use.

pfmoore commented:
@jkowalleck If we put the data into the core metadata (ignoring for now the problem that we don't have a viable way of doing so that conforms to the rules around dynamic and static metadata) , then we'd have to wait for a new metadata version to be implemented by build backends, recognised by PyPI and other tools, and supported throughout the ecosystem before it would be usable. The experience with the license-files metadata is that (a) this is a slow process, and (b) if tools try to implement support before the infrastructure is in place, we end up causing problems for users that take even *longer* to fix.

Conversely, if we add a new key to `pyproject.toml`, tools can add support for it with very little need to wait for infrastructure support. The `.dist-info` subdirectories are currently unregulated, so as long as tools write content that won't be invalidated by this PEP, they won't hit problems by being early in doing so.

But as @rgommers said, can this discussion be relocated to Discourse if you wish to continue it, please?

sethmlarson commented on peps/pep-0770.rst:
```diff
@@ -406,18 +350,69 @@ PyPI and other indices MAY validate the contents of SBOM documents specified by
+
+* Most subdirectories under ``.dist-info`` are to do with licensing,
+  one of which (``licenses``) is specified by :pep:`639` and other
+  (``license_files``) which is being used by the Hatch build backend.
```
@rgommers @ofek Apologies, I was misremembering which build backend had started building out this feature and what names they chose. After running a query to also inspect the `Generator` field in `WHEEL` for the archives which have `.dist-info` dirs I've found the following:

* Hatchling and PDM are created the `.dist-info/licenses` directories.
* Maturin and bdist_wheel created the `.dist-info/license_files` directories.
* Poetry created the `.dist-info/LICENSES` directories.
* cmeel created the `.dist-info/license` directories. This backend is only used by 3 projects (pin, eigenpy, and hpp_fcl).

For the `.dist-info/include` directory I found the following files:

* `include/pyobjc-compat.h`
* `include/pyobjc-api.h`

So only the pyobjc-core project is using this directory.

ofek replied:
The issue is one of eagerness rather than compliance: most drafts of PEP 639 until the end chose `license_files` rather than the final `licenses`.

rgommers replied:
`cmeel` is already fixed: https://github.com/cmake-wheel/cmeel/blob/8425feb36918eedb260ad1417ee41f6fa35f36da/cmeel/metadata.py#L37

So the way you have it now, allowing `license_files` and `LICENSES` for backwards compat, and forbidding the rest, looks good to me.

sethmlarson commented on peps/pep-0770.rst:
```diff
@@ -179,98 +179,58 @@ Specification
+
+Build backends MUST NOT create subdirectories in the ``.dist-info`` directory
+beyond the names in the registry to avoid collisions with future reserved names.
 
```
I adopted your suggestion of `MAY` validate, either with warnings or errors: https://github.com/python/peps/pull/4283/commits/a307eb43d5a73d060da7b986ddbfd1e49b37fbd1

sethmlarson commented:
Thanks @rgommers and @pfmoore for your review and kind words. I've pushed some changes (https://github.com/python/peps/pull/4283/commits/a307eb43d5a73d060da7b986ddbfd1e49b37fbd1) addressing feedback, please take a look.

sethmlarson commented on peps/pep-0770.rst:
```diff
@@ -613,8 +652,7 @@ Acknowledgements
 Thanks to Karolina Surma for authoring and leading :pep:`639` to acceptance.
-This PEP copies the specification from :pep:`639` for specifying files in
-project source metadata, Core Metadata, and project formats is based on.
+This PEP's initial design was heavily based on 639
```
Finished acknowledgement: https://github.com/python/peps/pull/4283/commits/037d2fbbe3a9824eab7477634c8a5fc253c78ed9

AA-Turner commented on peps/pep-0770.rst:
```diff
@@ -139,6 +139,29 @@ These decisions mean this PEP is capable of supporting any SBOM standard
+can projects begin using the new features. Even with this careful approach it's
+not guaranteed that tools won't break on new metadata versions and fields.
+
+To avoid this delay, this PEP proposes a new mechanism ``[additional-files]``
+to safely add data to a Python package through a registry of reserved names that
```
This might be more suited to a discuss.p.o comment, but `[additional-files]` feels a somewhat out of place to me as a name ("additional files for what?").

Something more self-describing that it is for built (or source) distributions, rather than for the source tree / development work might be nicer. For example `[dependency-groups]` is a top-level table that has no impact on distribution packages, so it isn't the case that top-level tables always relate to distribution packages.

The best I can come up with is `[distribution-files]` but it isn't great!

pfmoore replied:
The `additional-files` name was just my throwaway suggestion. I honestly expected way more bikeshedding than it got. But yes, this should probably be raised on Discourse.

AA-Turner replied:
See https://discuss.python.org/t/76308/76

AA-Turner commented on peps/pep-0770.rst:
```diff
@@ -179,98 +204,62 @@ Specification
-  `pyproject.toml specification <pyprojecttoml_>`__.
-* `Additions <770-spec-project-formats_>`_ to the source distribution (sdist),
-  built distribution (wheel), and installed project specifications.
+* A new reserved registry of subdirectory names in the ``.dist-info`` directory.
```
If I'm understanding correctly, the idea is that all names in `.dist-info` are reserved, and this registry is the ones that are allowed? If so, this wasn't entirely clear from the text, as I thought that some names might be unreserved. Perhaps?:
```suggestion
* Explicitly reserving all names  in the ``.dist-info`` directory.
* A new registry of reserved subdirectory names in the ``.dist-info`` directory.
```

AA-Turner commented on peps/pep-0770.rst:
```diff
@@ -179,98 +204,62 @@ Specification
-* `Additions <770-spec-project-formats_>`_ to the source distribution (sdist),
-  built distribution (wheel), and installed project specifications.
+* A new reserved registry of subdirectory names in the ``.dist-info`` directory.
+* A new reserved optional ``[additional-files]`` table with an optional
+  ``sboms`` key added to
+  `project source metadata <770-spec-project-source-metadata_>`_,
```
All top-level names in project source metadata are reserved. 
```suggestion
* An optional top-level table, ``[additional-files]``, added to
  `project source metadata <770-spec-project-source-metadata_>`_,
* An optional ``sboms`` key in the new ``[additional-files]`` table,
```

sethmlarson replied:
Done in https://github.com/python/peps/pull/4283/commits/f0258e8e81db992953ee2424c7af1600b34070cb

AA-Turner commented on peps/pep-0770.rst:
```diff
@@ -179,98 +204,62 @@ Specification
+and :term:`installed project` s project types. Future additions to this registry
+will be made through the PEP process. The initial values in this registry are:
+
+.. table::
```
You don't need the ``.. table::`` directive in reST if just using a simple table.

You could consider the `list-table` approach though:

```rst
.. list-table::
   :align: left
   :header-rows: 1

   * - Directory name
     - PEP
   * - ``licenses``
     -  :pep:`639`
   * - ``license_files``
     -  N/A (See :ref:`770-backwards-compat`)
   * - ``LICENSES``
     -  N/A (See :ref:`770-backwards-compat`)
   * - ``sboms``
     -  :pep:`770`
```

AA-Turner commented on peps/pep-0770.rst:
```diff
@@ -179,98 +204,62 @@ Specification
+    ``sboms``         :pep:`770`
+    ================= ==========
+
+Build backends MUST NOT create subdirectories in the ``.dist-info`` directory
+beyond the names in the registry to avoid collisions with future reserved names.
+
+Build frontends and publishing tools MAY warn users or raise a validation error
+if any ``.dist-info`` subdirectories aren't in the registry.
```
I would put this first, before the new registry, as marking all `.dist-info` subdirectories is distinct from the new registry. It may even deserve its own subsection.

sethmlarson replied:
I am not 100% sure I understood your comment, so can you double-check my changes in https://github.com/python/peps/pull/4283/commits/f0258e8e81db992953ee2424c7af1600b34070cb ?

AA-Turner replied:
All covered -- thanks!

AA-Turner commented on peps/pep-0770.rst:
```diff
@@ -406,18 +379,73 @@ PyPI and other indices MAY validate the contents of SBOM documents specified by
+being pervasively used and somehow load-bearing is unlikely, so is an accepted
+risk of using this method. This query yielded the following results:
+
+.. table::
```
Likewise, you don't need the `table` directive here, and you can de-dent the table itself

sethmlarson replied:
https://github.com/python/peps/pull/4283/commits/f0258e8e81db992953ee2424c7af1600b34070cb

brettcannon commented on peps/pep-0770.rst:
```diff
@@ -139,6 +139,29 @@ These decisions mean this PEP is capable of supporting any SBOM standard
+can projects begin using the new features. Even with this careful approach it's
+not guaranteed that tools won't break on new metadata versions and fields.
+
+To avoid this delay, this PEP proposes a new mechanism ``[additional-files]``
```
I think it's worth saying it isn't _just_ about getting this out faster as that feels like this is being done to cut a corner instead of doing what we think is the right approach.

```suggestion
To avoid this delay, simplify overall how to include SBOMs, and give flexibility to build back-ends, this PEP proposes a new mechanism ``[additional-files]``
```

sethmlarson replied:
Great idea, thanks all! I've adopted this phrasing in https://github.com/python/peps/pull/4283/commits/502496bb37912d55b844c7c00cdb2a5f70a7867f and included the reasoning for not using `[project]`.

AA-Turner commented on peps/pep-0770.rst:
```diff
@@ -406,13 +435,17 @@ Not shown above are around ~50 other subdirectory names that are used in a
-  (``license_files``) which is being used by the Hatch build backend.
+  one of which (``licenses``) is specified by :pep:`639` and others
+  (``license_files``, ``LICENSES``) are from draft implementations
+  of :pep:`639`.
```
#2164 (80e009085f9ab88f6555a60398601021e44a861a; 2021-12-17) mentions ``.dist-info/license_files``. This was changed to ``.dist-info/licenses`` in #2705 (d31806fe1e045e5fb2d92df38c510685f2595035; 2022-08-03), as noted in the commit message.

A

AA-Turner replied:
As such though, would it be better to say that tools supporting Metadata 2.5 MUST NOT use the `license_files` subdirectory? As part of supporting Metadata 2.4 tools need to use `licenses` instead, anyway.

It might also be nice to add the Core-Metadata version to the table of reserved `.dist-info` names.

rgommers replied:
That doesn't seem unreasonable, but it seems out of scope for this PEP to do that, since core metadata version is no longer bumped to 2.5. Who knows, maybe the next version is 3.0:)

> It might also be nice to add the Core-Metadata version to the table of reserved `.dist-info` names.

That version is recorded inside the `METADATA` file, it doesn't need a separate file or directory.

rgommers replied:
Looks great! All my comments have been addressed after that commit.

sethmlarson commented on peps/pep-0770.rst:
```diff
@@ -149,8 +172,10 @@ software within Python packaging metadata. This section aims to show how these
-  as a build-time dependency (``virtual:compiler/c``). PEP 770 describes
-  **concrete dependencies**, such as an exact name, version, architecture, and
+  as a build-time dependency (``virtual:compiler/c``) or an "OpenSSL-like
+  library" (``pkg:generic/openssl``). PEP 770 describes
```
Oh no reason in particular, I was reading too much into the `generic`! I've adopted your phrasing in this commit: https://github.com/python/peps/pull/4283/commits/502496bb37912d55b844c7c00cdb2a5f70a7867f

brettcannon commented:
FYI the latest changes make the title of the PR a bit misleading. 😉
